# (연습) 변수 할당

**참고**

먼저 [변수 할당](https://codingalzi.github.io/code-workout-python/starting/) 내용을 공부한 다음에 연습문제를
풀어볼 것을 권장한다.

## Hello, world! 출력 명령문

### 연습문제

아래 명령문들을 실행해서 오류가 발생하는 경우 오류의 종류와 원인을 설명하고 수정하라.
오류가 발생하지 않으면 실행 결과를 설명한다.

(1) 오류 수정 또는 실행 결과 설명

In [1]:
print("Hello World)

SyntaxError: unterminated string literal (detected at line 1) (2771583741.py, line 1)

답:

- 오류 종류: `SyntaxError`. 구문 오류
- 오류 발생 원인: `print()` 함수의 인자로 사용된 문자열을 감싸는 큰따옴표가 시작 부분에만 사용되었음.

문자열을 닫는 큰따옴표를 추가하면 오류가 발생되지 않는다.

In [2]:
print("Hello World")

Hello World


(2) 오류 수정 또는 실행 결과 설명

In [3]:
print(21 8)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (1287684157.py, line 1)

답:

- 오류 종류: `SyntaxError`. 구문 오류
- 오류 발생 원인: `print()` 함수의 인자가 이상함. 
    예를 들어, `21 8` 또는 `21+8`을 잘못 적었을 가능성 있음.
    21과 8 사이에 공백이 있으면 어떤 의미인지 알 수 없음.

예을 들어 21+8을 잘못 적은 거라면 아래처럼 수정하면 해결된다.

In [4]:
print(21+8)

29


(3) 오류 수정 또는 실행 결과 설명

In [5]:
print(2++2)

4


답:

오류가 발생하지 않는다. 
이유는 `+2`가 `2`로 간주되어 결국 `2 + 2`가 계산되어 4가 화면에 출력된다.

## 변수 할당문

### 연습문제

(1) 서점에 가서 15,000원짜리 책 한 권과 23,000원짜리 책 한 권을 샀다. 
이 서점은 현재 10% 할인을 행사 중이다. 
아래 단계에 따라 최종 결제 금액을 계산하고 출력하는 코드를 작성하라.

- `book1` 변수에 첫 번째 책 가격 15000 할당
- `book2` 변수에 두 번째 책 가격 23000 할당
- `total` 변수에 두 책의 합계 금액 할당
- `discount_rate` 변수에 할인율 0.1 할당
- `final_price` 변수에 할인율이 적용된 최종 결제 금액 할당 (공식: `total - (total * discount_rate)`)
- 최종 금액 출력

In [6]:
book1 = 15000
book2 = 23000
total = book1 + book2
discount_rate = 0.1

final_price = total - (total * discount_rate)
print(final_price)

34200.0


답:

`34200.0` 이 출력된다.
참고: 총액 38000원에서 10%인 3800원을 뺀 금액이다.

(2) 현재 1달러(USD)는 1350원(KRW)이라고 가정하자.
보유하고 있는 100달러가 한화로 얼마인지 계산하여 `krw_money` 변수에 할당하고 출력하라.
(힌트: `usd_rate = 1350`, `usd_money = 100` 변수 활용)

In [7]:
usd_rate = 1350
usd_money = 100
krw_money = usd_money * usd_rate
print(krw_money)

135000


답:

`135000` 이 출력된다.

(3) "2시간 30분 15초"가 총 몇 초인지 계산하여 `total_seconds` 변수에 할당하고 출력하라.
(힌트: 1시간 = 60분, 1분 = 60초)

In [8]:
hour = 2
minute = 30
second = 15

total_seconds = (hour * 60 * 60) + (minute * 60) + second
print(total_seconds)

9015


답:

`9015` 가 출력된다.
참고: `2 * 3600 + 30 * 60 + 15 = 7200 + 1800 + 15 = 9015`

## 변수 작성법

### 연습문제

(1) 아래 명령문들을 실행해서 오류가 발생하는 경우 오류의 종류와 원인을 설명하고 수정하라.
오류가 발생하지 않으면 실행 결과를 설명한다.

가. 오류 발생 원인 또는 실행 결과 설명

In [9]:
x = 2;

답:

오류가 발생하지 않는다.
이유는 세미 콜론 `;` 기호가 파이썬에서는 굳이 필요 없지만 연속된 두 개의 명령문을 한 줄에 작성할 때 활용되기도 한다.
예를 들어 아래 코드에서처럼 다른 변수 할당 명령문을 이어서 작성할 수 있다.

In [10]:
x = 2; y = 5
print('x:', x)
print('y:', y)

x: 2
y: 5


세미 콜론 뒤어 다른 명령문이 오지 않으면 무시된다.
따라서 위 코드는 아래 코드와 동일한 일을 수행한다.

In [11]:
x = 2
print('x:', x)

x: 2


나. 오류 발생 원인 또는 실행 결과 설명

In [12]:
y = 3
z = x y

SyntaxError: invalid syntax (457711640.py, line 2)

답:

- 오류 종류: `SyntaxError`. 구문 오류
- 오류 발생 원인: `x y`는 의미가 없는 표현식임. 곱셈, 나눗셈 등 연산자가 생략되었을 가능성이 큼.

In [13]:
y = 3
z = x * y

다. 오류 발생 원인 또는 실행 결과 설명

In [14]:
even = 3
odd = 17
print(even + oddd)

NameError: name 'oddd' is not defined

답:

- 오류 종류: `NameError`. 이름 오류.
- 오류 발생 원인: 변수 `oddd`가 정의되어 있지 않음. 앞서 정의된 `odd`를 잘못 타이핑한 것으로 보임.

이미 선언된 `odd`를 `oddd` 대신 사용하면 오류가 발생하지 않는다.

In [15]:
even = 3 # 홀수 값
odd = 17 # 짝수 값
print(even + odd)

20


라. 오류 발생 원인 또는 실행 결과 설명

In [16]:
money! = 1000

SyntaxError: invalid syntax (1792621820.py, line 1)

답:

변수명에 `!` 같은 특수문자를 사용할 수 없음을 확인 (SyntaxError).

(2) 아래 낙타 표기법 변수를 뱀 표기법 변수로 대체하라.

In [17]:
TopScore = 100

답: 

파이썬에서 변수명의 첫 글자는 특별한 경우를 제외하고는
일반적으로 소문자로 시작하는 것이 관례다.
따라서 아래와 같이 뱀 표기법 변수가 권장된다.

In [18]:
top_score = 100

## 변수 재할당

### 연습문제

(1) 다음 코드의 실행 결과를 예측하라.

In [19]:
box = '사과'
box = '귤'
box = '배'
print(box)

배


답:

`'배'`가 출력된다. 변수 `box`에는 처음에 `'사과'`, 그 다음 `'귤'`, 마지막으로 `'배'`가 할당되었다.
`print(box)` 함수가 실행되는 시점에는 가장 마지막에 할당된 값인 `'배'`를 가리키고 있기 때문이다.

(2) 세 변수가 다음과 같이 주어졌다.

In [20]:
x = 5
y = 9
z = 2

아래 재할당을 순서대로 실행했을 때 세 `x`, `y`, `z`에 최종적으로 할당된 값을 예측하라.

In [21]:
x = y + z
y = x * 2
z = y - 3

답:

In [22]:
print('x:', x)
print('y:', y)
print('z:', z)

x: 11
y: 22
z: 19


## 변수 업데이트

### 연습문제

(1) 변수 `price`가 처음에는 1000이었다가, 10% 가격이 인상된 후의 값을 가지도록 재할당하는 코드를 작성하라.

(힌트: `price = price * ...` 형태 활용)

답:

In [23]:
price = 1000
price = price * 1.1
print(price)

1100.0


(2) 두 변수가 다음과 같이 주어졌다.

In [24]:
x = 10
y = 3

다음 코드가 실행된 뒤 `x`, `y`가 최종적으로 가리키는 값을 예측하라.

In [25]:
x %= y
y += x
x **= 2
y -= 1

답:

In [26]:
print('x:', x)
print('y:', y)

x: 1
y: 3


## 자료형 확인

### 연습문제

(1) 아래 코드의 실행 결과를 설명하라.

In [27]:
5.2 * 7

36.4

답:

부동소수점과 소수점의 연산은 부동소수점으로 계산된다.

(2) 아래 표현식을 실행하면 오류가 발생하는 이유를 설명하라.

In [28]:
2.0 * "7.18"

TypeError: can't multiply sequence by non-int of type 'float'

답:

- 오류 종류: `TypeError`. 자료형 오류.
- 오류 발생 원인: 부동소수점과 문자열의 곱셈은 허용되지 않음.

부동소수점 2.0와 정수 2가 일반적으로는 그 기능이 다름을 잘 보여준다.
예를 들어 2.0 대신 2를 사용하면 오류가 발생하지 않는다

In [29]:
2 * "7.18"

'7.187.18'

연산자와 연산에 사용되는 값의 자료형에 따라 연산의 허용 여부가 결정된다는 점을 기억해 두어야 한다.

(3) 정수 10과 부동소수점 2.0을 더한 결과의 자료형이 무엇인지 `type()` 함수를 사용하여 확인하고 설명하라.

답:

In [30]:
type(10 + 2.0)

float

정수와 부동소수점의 합은 부동소수점으로 계산된다.
이유는 더해지는 두 값의 자료형을 통일시키기 위해 10을 10.0으로 변환시킨 다음에
10.0과 2.0을 더하기 때문이다.

## 파이썬 계산기

### 연습문제

(1) 먼저 아래와 같이 두 개의 변수를 정의한다.

In [31]:
width = 17
height = 12.0

두 변수를 이용하여 아래에 지정된 값을 가리키는 표현식을 작성한 다음에 실행 결과를 확인하라.

가. width를 3으로 나눈 나머지

답:

17을 3으로 나누면 몫은 5, 나머지는 2다. 

In [43]:
width % 3

2

나. width 나누기 3.0

답:

나눗셈은 무조건 부동소수점으로 계산된다.

In [44]:
width / 3.0

5.666666666666667

다. height 빼기 3

답:

사칙연산에 부동소수점이 포함되면 무조건 부동소수점으로 계산된다.

In [45]:
height - 3

9.0

라. height 곱하기 5에 7 더하기

답:

사칙연산에 부동소수점이 포함되면 무조건 부동소수점으로 계산된다.

In [46]:
height * 5 + 7

67.0

마. height 곱하기 5에 7.0 더하기

사칙연산에 부동소수점이 포함되면 무조건 부동소수점으로 계산된다.

답:

In [47]:
height * 5 + 7.0

67.0

(2) 오전 6시 52분에 집을 떠나서,
초반에는 km당 8분 15초의 속력으로 1km를 뛰고, 
이어서 km당 7분 12초의 속력으로 3km를 뛴 후, 
초반 속력으로 1km를 마저 달려 집으로 돌아왔을 때의 도착시간을 계산하는 코드를 작성하라.
단 집에 도착하는 시간을 아래와같이 출력되도록 해야 한다.

```
집에 도착 시간: 7 시 30 분 6 초
```

힌트: 시, 분, 초를 동시에 다루기 위해 계산을 초 단위로 통일시킨 후 아래 변수들을 정의하고 최종적으로 도착 시간을 지정된 형식으로 출력한다.

- `start`: 자정 기준 출발 시간까지 지난 시간
- `running_time`: 달린 시간
- `arrival_time`: 자정 기준 도착 시간까지 지난 시간

답:

가. 초단위 도착시간 계산

시, 분, 초를 별도로 다루면 계산이 복잡해지기에 모든 값을 초 단위로 통일시킨다.
먼저 출발시간인 6시 52분이 밤 0시를 기준으로 몇 초 지난 시간인지
계산해서 `start` 변수에 할당한다.

In [32]:
start = (6 * 60 + 52) * 60
print(start)

24720


아래 두 변수는 1 킬로미터를 가볍게 달릴 때와 좀 빠르게 달릴 때 걸리는 시간(초)을 가리킨다.

In [33]:
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

그러면 달리는 시간을 초로 계산하면 2286초로 나온다.

In [34]:
running_time = slow_running + 3 * fast_running + slow_running
running_time

2286

도착시간은 출발시간에 달린 시간을 더하면 된다.

In [35]:
arrival = start + running_time
arrival

27006

나. 시, 분, 초로 변환

이제 도착시간을 시, 분, 초로 변환하기 위해 1분은 60초, 1시간은 60분이라는 정보를
이용하여 도착시간에 대해 60으로 나눴을 때의 몫과 나머지 계산을 두 번 연속 적용한다.

- 도착시간의 초 계산: 60으로 나눴을 때의 나머지

In [36]:
arrival_seconds = arrival % 60
arrival_seconds

6

- 도착시간의 분 계산: 60으로 나눴을 때의 몫

In [37]:
arrival_minutes = arrival // 60
arrival_minutes

450

450분은 60분 보다 크기에 시와 분으로 다시 변환해야 한다.
한 시간이 60분이기에 역시 60으로 나눈 몫과 나머지를 계산한다.

- 도착시간의 시 계산: 60으로 나눴을 때의 몫

In [38]:
arrival_hours = arrival_minutes // 60
arrival_hours

7

- 도착시간의 분 계산: 60으로 나눴을 때의 나머지

In [39]:
arrival_minutes = arrival_minutes % 60
arrival_minutes

30

따라서 도착시간은 7시 30분 6초가 된다. 

In [40]:
print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초


다. 종합 정리

지금까지 설명한 모든 코드를 종합하면 다음과 같다.

In [41]:
start = (6 * 60 + 52) * 60
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

running_time = (1 + 1) * slow_running + 3 * fast_running
arrival = start + running_time

arrival_seconds = arrival % 60
arrival_minutes = arrival // 60
arrival_hours = arrival_minutes // 60
arrival_minutes = arrival_minutes % 60

print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초


(3) 100만원을 복리로 연이자 5%인 예금에 3년간 은행에 맡겼을 때
3년 후에 수령할 금액을 계산하는 코드를 다양한 변수를 활용하여 구현하라.
단, 다음 형식으로 출력한다.

```
3년 후 받을 원금 + 이자는 OOO 원입니다.
```

복리계산은 아래 식을 따른다.

```
원리합계 ＝ 원금 * (1 ＋ 연금리/일년에 이자받는횟수) ** (일년에 이자받는횟수 * 기간)
```

아래 변수를 활용한다.

- `principal`: 원금
- `rate`: 연이자율
- `frequency`: 이자 지급 횟수 (연단위)
- `period`: 저축기간 (년)
- `savings`: 원리합계

힌트: [복리 개념과 계산법](https://a2-finance.com/ko/calculators/%EB%AA%A8%EB%93%A0-%EA%B3%84%EC%82%B0%EA%B8%B0/%EB%B3%B5%EB%A6%AC-%EA%B3%84%EC%82%B0%EA%B8%B0)
참고

답

정해진 변수와 원리합계 식을 주어진대로 반영하여 코드를 작성하면 된다.

In [50]:
# 원금
principal = 1_000_000 # 밑줄 기호는 천단위 숫자 구분자로 사용됨. 
# 연이자율
rate = 0.05
# 이자 지급 횟수 (연단위)
frequency = 1
# 저축기간 (년)
period = 3
# 원리합계
savings = principal * (1 + rate/frequency) ** (frequency * period)

print("3년 후 받을 원금 + 이자는", savings, "원입니다.")

3년 후 받을 원금 + 이자는 1157625.0000000002 원입니다.
